### PR TITLE
Add/remove interfaces to zone using PCI device ID

### PR DESCRIPTION
**Pull request by BrennanPaciorek**
- Add/remove interfaces to a zone using PCI ID (formatted XXXX:XXXX, VendorID:DeviceID)
  - Collision of interfaces with the same PCI ID (while unlikely) is handled by applying the play to each interface with the PCI id
  - fixes RHBZ 2100939
  - Useful as it allows control nodes to manage interfaces by what a device is rather than what interface name it was assigned.
- Added Network Manger interfacing in the same capacity to which firewall-cmd interacts with Network Manager for interfaces
  - See `--[add/remove/change]-interface`  at https://firewalld.org/documentation/man-pages/firewall-cmd.html

### CODE REVIEW

**Comment by BrennanPaciorek** [comment]
I'm still tweaking the test, as there will most likely be issues when dealing with traditional interface names and systemd predictable interface names

File: library/firewall_lib.py [hunk 1]
```diff
@@ -323,28 +311,33 @@ def handle_interface_permanent(
...
         device_path = os.path.join(device_udi, "device")
         for field in ["vendor", "device"]:
             with open(os.path.join(device_path, field)) as _file:
                 interface_ids.append(_file.readline().strip(" \n")[2:])
-
-        pci_dict[":".join(interface_ids)] = interface
+        interface_ids = ":".join(interface_ids)
+        if interface_ids not in pci_dict:
+            pci_dict[interface_ids] = [interface]
+        else:
```
Comment (liangwen12year) [inline library/firewall_lib.py:333]: Maybe detailed commit message is needed for explaining why this commit is needed.

File: README.md [hunk 2]
```diff
@@ -199,10 +199,18 @@ source: 192.0.2.0/24
 ### interface
 
 String or list of interface name strings.
+Also accepts PCI id's, which is two pairs of hexidecminal numbers (XXXX:XXXX)
```
Comment (liangwen12year) [inline README.md:202]: If the regex applies to the PCI id matching ?

File: README.md [hunk 3]
```diff
@@ -199,10 +199,18 @@ source: 192.0.2.0/24
...
-```
+If a PCI id corresponds to more than one interface name,
+all interfaces with the PCI id specified will have the play applied.
+
+```yaml
 interface: eth2
 ```
+```yaml
+# PCI id for Intel Corporation Ethernet Connection
+interface: 8086:15d7
```
Comment (liangwen12year) [inline README.md:212]: FYI, in the network role, we use a brand new key name (`path`) to specify the PCI address [1], not sure if the firewall role should also have a good key name distinction between the actual interface name and PCI ID.

[1] https://github.com/linux-system-roles/network/blob/main/module_utils/network_lsr/argument_validator.py#L1755

Comment (liangwen12year) [reply README.md:212]: Also adding an complete vivid example like the following would be nice:

Allow interface with PCI ID 8086:15d7  in trusted zone:

```yaml
firewall:
  - interface: 8086:15d7
    zone: trusted
    state: enabled
```

File: library/firewall_lib.py [hunk 4]
```diff
@@ -222,6 +224,48 @@
...
+    from firewall.core.fw_nm import (
+        nm_is_imported,
+        nm_get_connection_of_interface,
+        nm_get_zone_of_connection,
+        nm_set_zone_of_connection,
+        nm_get_interfaces,
+        nm_get_client,
+    )
+
+    NM_IMPORTED = nm_is_imported()
```
Comment (vrindle) [inline library/firewall_lib.py:237]: Is there a benefit to calling this function rather than just setting it to true?

File: library/firewall_lib.py [hunk 5]
```diff
@@ -222,6 +224,48 @@
...
+
+# Below: Adapted from firewall-cmd source code
+def try_set_zone_of_interface(module, _zone, interface):
+    if NM_IMPORTED:
+        try:
+            connection = nm_get_connection_of_interface(interface)
+        except Exception:
+            pass
+        else:
+            if connection is not None:
```
Comment (vrindle) [inline library/firewall_lib.py:250]: Very nested. This should all probably be in its own function. With so many conditionals it can be hard to follow.

Comment (BrennanPaciorek) [reply library/firewall_lib.py:250]: I'll try to make it more readable

File: library/firewall_lib.py (hunk 4 above)
Comment (BrennanPaciorek) [reply library/firewall_lib.py:237]: The function actually checks if Network Manager itself is imported in fw_nm.py, which is not checked by importing the functions. If the import fails, we can just assume no, but if it succeeds, the function call here I imagine would be better than calling `nm_is_imported()` for each zone changed.
https://github.com/firewalld/firewalld/blob/master/src/firewall/core/fw_nm.py#L31

File: README.md (hunk 3 above)
Comment (BrennanPaciorek) [reply README.md:212]: After thinking about it, I agree on both comments on this conversation, it would be difficult to manage an interface named like a PCI device ID without making the change you suggested or a similar change, although I have no idea why someone would knowingly add an interface with the name in the structure of a PCI id.

Additionally, I will go ahead and add an example to interface using a PCI device ID after making the first suggested change

File: library/firewall_lib.py (hunk 1 above)
Comment (BrennanPaciorek) [reply library/firewall_lib.py:333]: I updated the PR summary, does that suffice?

Comment (liangwen12year) [reply library/firewall_lib.py:333]: Well, here is a good article about how to write the commit message, https://cbea.ms/git-commit/, you can take a look.

Comment (richm) [reply library/firewall_lib.py:333]: The project has a git commit message section in the contribution guide https://linux-system-roles.github.io/contribute.html
"Write a good commit message"

File: library/firewall_lib.py [hunk 6]
```diff
@@ -269,6 +323,39 @@ def handle_interface_permanent(
...
+
+def parse_pci_id(module, item):
+    if PCI_REGEX.search(item):
+        global pci_ids
+        if not pci_ids:
+            pci_ids = get_interface_pci()
+        interface_name = pci_ids.get(item)
+        if interface_name:
+            return interface_name
+    module.fail_json(msg="No network interfaces found with PCI device ID %s" % item)
```
Comment (BrennanPaciorek) [inline library/firewall_lib.py:356]: TODO: error if PCI id is invalid, rather than a single fail message in parse_pci_id. Unit test for both errors

**Review by BrennanPaciorek — COMMENTED** [review]
TODO:
- Integration testing: needs to support case that iptables is the backend, meaning need a good way to get firewalld backend. firewalld.conf seems to not be present on installations of some VMs or is in unusual location.

File: library/firewall_lib.py [hunk 7]
```diff
@@ -269,6 +323,44 @@ def handle_interface_permanent(
...
+
+def parse_pci_id(module, item):
+    if PCI_REGEX.search(item):
+        global pci_ids
+        if not pci_ids:
+            pci_ids = get_interface_pci()
+        interface_name = pci_ids.get(item)
+        if interface_name:
+            return interface_name
+        module.fail_json(msg="No network interfaces found with PCI device ID %s" % item)
```
Comment (BrennanPaciorek) [inline library/firewall_lib.py:356]: I need to change this to a warning, module shouldn't fail because someone tries to add an interface that happens to not exist on the managed node.

File: tests/tests_interface_pci.yml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        string:"org.fedoraproject.FirewallD1.config"
+        string:"FirewallBackend"
+      ignore_errors: yes
+      register: result
+
+    - name: get backend from result
+      set_fact:
+        firewall_backend:
+          "{{ 'nftables' in result.stdout |
+           ternary('iptables', 'nftables', 'iptables') }}"
```
[the remark this answers is not in the document]
Comment (BrennanPaciorek) [reply tests/tests_interface_pci.yml:23]: Oh right, we can't be sure this block would work without calling the firewall role first. The ansible fact PR (https://github.com/linux-system-roles/firewall/pull/83) currently uses up the functionality of calling the role with no options. That PR may need to be revised to preserve the functionality of people just looking to install firewalld dependencies and enable the service remotely.

File: library/firewall_lib.py [hunk 9]
```diff
@@ -447,6 +539,7 @@ def main():
             rich_rule=dict(required=False, type="list", elements="str", default=[]),
             source=dict(required=False, type="list", elements="str", default=[]),
             interface=dict(required=False, type="list", elements="str", default=[]),
+            interface_pci=dict(required=False, type="list", elements="str", default=[]),
```
[the remark this answers is not in the document]
Comment (vrindle) [reply library/firewall_lib.py:546]: `interface_id` is a good name but I think we could be more explicit and call it something like `interface_pci_id`. What do you think? @BrennanPaciorek @liangwen12year

File: library/firewall_lib.py (hunk 1 above)
Comment (vrindle) [reply library/firewall_lib.py:344]: @richm @liangwen12year I see there's an updated commit message, it looks good. What do you folks think?

File: library/firewall_lib.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (BrennanPaciorek) [reply library/firewall_lib.py:546]: This sounds good to me, I'll make the changes Monday morning.

**Comment by richm** [comment]
[citest commit:ed1b763a125d77dec38ef56204c30e6fa542e9dc]

File: library/firewall_lib.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (liangwen12year) [reply library/firewall_lib.py]: `interface_pci_id` works for me

File: library/firewall_lib.py (hunk 1 above)
Comment (liangwen12year) [reply library/firewall_lib.py:344]: The updated commit message lgtm.

**Comment by richm** [comment]
[citest commit:ed1b763a125d77dec38ef56204c30e6fa542e9dc]

**Comment by BrennanPaciorek** [comment]
I need to update the test for this feature, i need to do some checking for dbus errors in case there is no "stdout" when fetching FIREWALL_BACKEND

File: README.md [hunk 10]
```diff
@@ -200,10 +200,33 @@ source: 192.0.2.0/24
...
+```yaml
 interface: eth2
 ```
 
+### interface_pci_id
+
+String or list of interface PCI device IDs.
+Accepts PCI IDs if the wildcard `XXXX:YYYY` applies
+where:
+- XXXX: Hexidecimal, corresponds to Vendor ID
```
Comment (richm) [inline README.md:212]: Hexadecimal

File: library/firewall_lib.py [hunk 11]
```diff
@@ -269,6 +324,47 @@ def handle_interface_permanent(
             fw_settings.addInterface(item)
 
 
+pci_ids = None
+
+
+def get_interface_pci():
+    pci_dict = {}
+    interfaces = nm_get_interfaces()
+    for interface in interfaces:
```
Comment (richm) [inline library/firewall_lib.py:333]: you don't need `interfaces` - `for interface in nm_get_interfaces():`

File: tests/tests_interface_pci.yml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+              zone: internal
+              interface_pci_id: 1af4:0001
+              state: enabled
+              permanent: yes
+
+        - name: assert pcid not in nftable ruleset
+          command: nft list ruleset
+          register: result
+          failed_when: "result.rc != 0 or '1af4:0001' in result.stdout"
+          when: nftables_backend
```
Comment (richm) [inline tests/tests_interface_pci.yml:51]: I think you need to use `nftables_backend | bool` here to avoid the "bare variable" problem

File: tests/tests_interface_pci.yml [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+            firewall:
+              zone: internal
+              interface_pci_id: 1af4:0001
+              state: enabled
+              permanent: yes
+
+        - name: assert pcid not in nftable ruleset
+          command: nft list ruleset
+          register: result
+          failed_when: "result.rc != 0 or '1af4:0001' in result.stdout"
```
Comment (richm) [inline tests/tests_interface_pci.yml:50]: `failed_when: result is failed or '1af4:0001' in result.stdout`

File: tests/tests_interface_pci.yml [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+          when: nftables_backend
+
+        - name: assert pcid not in iptables rules
+          command: iptables -S
+          register: result
+          failed_when: "result.rc != 0 or '1af4:0001' in result.stdout"
```
Comment (richm) [inline tests/tests_interface_pci.yml:56]: `failed_when: result is failed or '1af4:0001' in result.stdout`

File: tests/unit/test_firewall_lib.py [hunk 15]
```diff
@@ -360,6 +374,24 @@ def test_parse_forward_port(self):
...
+    @patch("firewall_lib.HAS_FIREWALLD", True)
+    @patch("firewall_lib.pci_ids", {"600D:7C1D": ["eth0"]})
+    def test_parse_pci_id(self, am_class):
+        am = am_class.return_value
+
+        am.params = {"interface_pci_id": ["123G:1111"]}
+        with self.assertRaises(MockException):
+            firewall_lib.main()
+        am.fail_json.assert_called_with(
+            msg="PCI id 123G:1111 does not match format: XXXX:XXXX (X = hexidecimal number)"
```
Comment (richm) [inline tests/unit/test_firewall_lib.py:387]: hexadecimal

File: library/firewall_lib.py [hunk 16]
```diff
@@ -269,6 +324,47 @@ def handle_interface_permanent(
...
+            pci_ids = get_interface_pci()
+
+        interface_name = pci_ids.get(item)
+        if interface_name:
+            return interface_name
+
+        module.warn(msg="No network interfaces found with PCI device ID %s" % item)
+    else:
+        module.fail_json(
+            msg="PCI id %s does not match format: XXXX:XXXX (X = hexidecimal number)"
```
Comment (richm) [inline library/firewall_lib.py:362]: hexadecimal

**Comment by BrennanPaciorek** [comment]
Okay that should fix the testsuite, CI tests just need to be run to be sure. All feedback has been implemented in the same update to the branch

**Comment by BrennanPaciorek** [comment]
[citest commit:eb1d5b1e41b20cb7cf808ded3e2aedad1653e24f]

**Comment by richm** [comment]
[citest]

**Comment by richm** [comment]
ok - lgtm - @vrindle @liangwen12year ?

File: README.md [hunk 17]
```diff
@@ -420,6 +443,22 @@ firewall:
     state: disabled
 ```
 
+Allow interface named '8086:15d7' in dmz zone
```
Comment (briansmith0) [inline README.md:446]: Not sure I understand this example -- are there circumstances where the interface name would be a PCI device ID?

Comment (BrennanPaciorek) [reply README.md:446]: Firewalld supports adding interfaces by any name, if it doesn't match with any logical interface name, it still gets added as a firewall rule without going through NetworkManager, but it likely will not match with any traffic. Since firewalld supports this, I felt it would be better practice to support it as well, even if it is just for removing interface names that do not match up with any network traffic.

Comment (briansmith0) [reply README.md:446]: Thanks for the context.  I would suggest that we mention this info you provided under the "interface" section of the README and remove this particular example from the README as it sounds like this example is not something that would generally be used or recommended, and the example might cause some confusion (it did for me at least).  What are your thoughts on this?

Comment (BrennanPaciorek) [reply README.md:446]: I agree, if it mislead you, I should change how the information is presented. I will move this information over to the interface section as a warning, since it is something the general user would need to know about when adding/moving interfaces.

Comment (BrennanPaciorek) [reply README.md]: Updated README, should the new warning under `interface` be visible enough to users?

Comment (briansmith0) [reply README.md]: This looks good to me and is very easy to understand.  Thanks!

**Comment by richm** [comment]
ok - I think this is ready to merge, unless there are any other objections?

**Comment by vrindle** [comment]
LGTM
